docs/source/more_types.rst

```diff
@@ -178,6 +178,211 @@ create subclasses of these objects either.
+        if len(components) == 4:
+            # Return an IPv4Address object
+        elif len(components) == 8:
+            # Return an IPv6Address object
+        else:
+            # Raise an exception
+
+This allows mypy to understand calls to ``ip_address`` much more precisely.
+For example, mypy will understand that ``ip_address(127, 0, 0, 1)`` will
+always have a return type of ``IPv4Address`` and will report errors for
+calls like ``ip_address(1, 2)``.
+
+As another example, suppose we want to write a custom container class that
+implements the ``__getitem__`` method (``[]`` bracket indexing). If this
+method receives an integer we return a single item. If it receives a
+``slice``, we return a ``Sequence`` of items.
+
+We can precisely encode this relationship between the argument and the
+return type by using overloads like so:
+
+.. code-block:: python
+
+    from typing import Sequence, TypeVar, Union
+
+    T = TypeVar('T')
+
+    class MyList(Sequence[T]):
+        @overload
+        def __getitem__(self, index: int) -> T: pass
+
+        @overload
+        def __getitem__(self, index: slice) -> Sequence[T]: pass
+
+        def __getitem__(self, index: Union[int, slice]) -> Union[T, Sequence[T]]:
+            if isinstance(index, int):
+                # Return a T here
+            elif isinstance(index, slice):
+                # Return a sequence of Ts here
+            else:
+                raise TypeError(...)
+
+There are a few additional things to note about using overloads:
+
+Runtime behavior
+----------------
+
+An overloaded function must consist of two or more overload *variants*
+followed by an *implementation*. The variants and the implementations
+must be adjacent in the code: think of them as one indivisible unit.
+
+The variant bodies must all be empty; only the implementation is allowed
+to contain code. This is because at runtime, the variants are completely
+ignored: they're overridden by the final implementation function.
+
+This means that an overloaded function is still an ordinary Python
+function! There is no automatic dispatch handling: you must manually
+handle the different types in the implementation (usually by using
+``if`` statements and ``isinstance`` checks).
+
+If you are adding an overload within a stub file, the implementation
+function should be omitted: stubs do not contain runtime logic.
+
+Type checking calls to overloads
+--------------------------------
+
+When you call an overloaded function, mypy will infer the correct
+return type using the provided variants. A call is never type checked
+against the implementation signature. This is why mypy will report calls
+like ``ip_address(4)`` as being invalid even though it matches the
+implementation signature.
+
+If multiple variants end up matching a call, mypy will, for the
+most part, select the return type corresponding to the first
+matching call. For example, consider the following program:
```

Comment:
I would reformulate this slightly. We can say something like: "Mypy will chose the best matching variant (considering argument kinds and types). If there are more than one equally good match, mypy will choose the first one. An exception is when there are multiple matching due to an `Any` type of some of the arguments, in this case mypy will infer `Any` as the result of the call."

Reply:
> Is there still a gradient in how "good" a match is?

Yes, we still use some heuristics, we discussed for some time and agreed it would be impossible to specify all corner cases in a way making everyone happy. So we deliberately leave some room for type checkers in ambiguous corner cases (the cases are quite well specified, but the behavior in each is up to type checker).

Also, @Michael0x2a how about you prepare a PR to peps repo using your original proposal on typing tracking and taking into account the things you learned from implementing all this?

> And what about provided context? Does that play a role?

Currently it is used only if there type variables I think. I think in future we can use it to pick up s better overload, but it is more complex that it seems.

Reply:
> the cases are quite well specified

Remind me where? PEP 484 explains almost entirely through examples. And the mypy docs don't apply to other type checkers. Are the heuristics only relevant to Protocols (PEP 544)? Then that should be called out specifically.

Reply:
> Remind me where?

Original PEP 484 amendment proposal https://github.com/python/typing/issues/253#issuecomment-389262904

Some amendments to amendment discussion is also here https://github.com/python/mypy/issues/5119

> Are the heuristics only relevant to Protocols (PEP 544)?

No, there are several other corner cases: empty collections, star args, union types.

Reply:
Also beware, the original proposal is _very_ long. It also discusses motivation, but the PEP 484 PR should include mostly the specification part IMO.

Reply:
Possibly it can become a separate PEP, and PEP 484 should be amended with a link to the new PEP? That shouldn't be a big deal (it's not like type checking PEPs seem to attract a lot of attention from the usual crowd of armchair critics).

Reply:
> Possibly it can become a separate PEP, and PEP 484 should be amended with a link to the new PEP?

I have no preferences, I original thought it would be OK to add this to the PEP 484 because it already contains a section on overloads, but having another mini-PEP is also OK.

Reply:
I expect evolving and reviewing the doc as a separate PEP will be easier. Otherwise it'll be a *very* large expansion to an already-approved PEP, and that doesn't sound completely fair to people who reviewed it originally.